downstairs/src/lib.rs

```diff
@@ -1226,14 +1241,46 @@ where
                         gen,
                         read_only,
                         encrypted,
+                        supported_versions,
                     }) => {
                         if negotiated != 0 {
                             bail!("Received connect out of order {}",
                                 negotiated);
                         }
-
-                        if version != 1 {
-                            bail!("expected version 1, got {}", version);
+                        info!(log, "Connection request from {} with version {}",
+                            upstairs_id, version);
+
+                        // Verify we can communicate with the upstairs.  First
+                        // check our message version.  If that fails,  check
+                        // to see if our version is one of the supported
+                        // versions the upstairs has told us it can support.
+                        if version != CRUCIBLE_MESSAGE_VERSION {
+                            if supported_versions
+                                .contains(&CRUCIBLE_MESSAGE_VERSION)
+                            {
+                                warn!(
+                                    log,
+                                    "downstairs requires down rev version {}",
+                                    CRUCIBLE_MESSAGE_VERSION
+                                );
+                            } else {
+                                let m = Message::VersionMismatch {
+                                    version: CRUCIBLE_MESSAGE_VERSION,
+                                };
+                                let mut fw = fw.lock().await;
+                                if let Err(e) = fw.send(m).await {
+                                    warn!(
+                                        log,
+                                        "Failed to send VersionMismatch: {}",
+                                        e
+                                    );
+                                }
+                                bail!(
+                                    "Required version {}, got {}",
```

Comment:
this should list the supported versions that the Upstairs sent too

Reply:
+1